File: mcomp_Java/src/member/coms/Pipe.java
package member.coms;

import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;
import common.datatypes.Waypoint;
import common.datatypes.map.MapLayer;
import common.objects.Member;
// serial communication library import
import jssc.SerialPort;
import jssc.SerialPortException;
import jssc.SerialPortList;

/**
 * A communication class for handling transmission and receiving of data over a serial connection
 * between the Pi and the Arduino.
 * 
 * @author Ryan Shoobert 15812407
 * @author David Avery 15823926
 * 
 * @version 2.1
 * @since 2018-05-10
 * 
 */
public class Pipe {
  // global definitions
  private SerialPort p;

  String[] availablePorts;

  // Connection Parameters - Should never change these once Pipe instance created
  // assume default baud rate for Arduino board, eight data bits, two stop bits, 1 bit for
  // parity
  private final int BAUD_RATE = 115200;
  private final int NUM_OF_DATA_BITS = 8;
  private final int NUM_OF_STOP_BITS = 1;
  private final int NUM_OF_PARITY_BITS = 0;

  private static final Logger LOGGER = Logger.getLogger(Pipe.class.getName());

  /**
   * Constructor for the Pipe Class. This will be responsible for opening a Serial connection on the
   * defined port and then configuring the connection parameters to those defined.
   * 
   * @param comPortName The COM port that the connection will be set up on
   */
  public Pipe() {
    availablePorts = SerialPortList.getPortNames();

    if (availablePorts.length == 0) {
      LOGGER.log(Level.SEVERE, "There are no serial-ports available");
    }

    // Print out list of available ports
    LOGGER.log(Level.INFO, "Available Ports:");
    for (int i = 0; i < availablePorts.length; i++) {
      LOGGER.log(Level.INFO, availablePorts[i]);
    }

    p = new SerialPort(availablePorts[0]);
    LOGGER.log(Level.INFO, "Using Port: " + availablePorts[0]);

    try {
      p.openPort();
      p.setParams(BAUD_RATE, NUM_OF_DATA_BITS, NUM_OF_STOP_BITS, NUM_OF_PARITY_BITS);
    } catch (SerialPortException e) {
      e.printStackTrace();
    }
  }

  /**
   * Builds a command and sends it. This instructs the driver on the robot to drive to the x, y
   * location provided in a waypoint.
   * 
   * @param w The waypoint to drive to
   * @return The waypoint returned from the drive call. This will either be waypoint it was told to
   *         drive to or where the robot got up to before an obstacle was detected.
   */
  public Waypoint drive(Waypoint w) {
    Waypoint res = new Waypoint(0, 0);

    try {
      res = decodeWaypoint(call(encode(Commands.DRIVE, w)));
    } catch (SerialPortException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

    return res;
  }

  /**
   * As a result of a drive operation, a waypoint is returned from the driver as a string. This will
   * require separating into parts and will be used to form a new waypoint.
   * 
   * @param s The waypoint data returned from the driver
   * @return A waypoint containing the x, y coordinates of where the robot was at the end of the
   *         drive operation
   */
  private Waypoint decodeWaypoint(String s) {
    Waypoint res = new Waypoint(0, 0);
    String[] xy = s.split(";");

    res = new Waypoint(Double.parseDouble(xy[0]), Double.parseDouble(xy[1]));
    return res;
  }

  /**
   * Requests a LiDAR return from the sensor, turns it into a map layer and returns it
   * 
   * @return A Map Layer representing the lidar return just collected
   */
  public MapLayer lSense() {
    LOGGER.log(Level.INFO, "begin requesting lread");

    MapLayer res = new MapLayer(null); // or new ArrayList<Waypoint>()

    try {
      res = decodeLSense(call(encode(Commands.L_SENSE, null)));
    } catch (SerialPortException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

    return res;
  }

  /**
   * Once requested, a lidar reading will be taken and returned. This method will separate each part
   * of that return into a waypoint, add it to a new map layer and return it where it can then be
   * added to the map.
   * 
   * @param s The lidar reading returned from the sense operation
   * @return A map layer representing the lidar read taken
   */
  private MapLayer decodeLSense(String s) {
    ArrayList<Waypoint> layer = new ArrayList<Waypoint>();
    String[] xys = s.split(";");

    for (int i = 0; i <= (xys.length - 2); i = i + 2) {
      layer.add(new Waypoint(Double.parseDouble(xys[i]), Double.parseDouble(xys[i + 1])));
    }

    MapLayer res = new MapLayer(layer);
    return res;
  }

  /**
   * A member needs to know the angle of direction it is facing in when it starts up so that when it
   * takes lidar reads, these can be rotated by the correct angle and amalgamated to the map.
   * 
   * @return The reading from the compass on the calling member
   */
  public double compass() {
    double res = 0.0;

    try {
      res = decodeDouble(call(encode(Commands.COMPASS, null)));
    } catch (SerialPortException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

    return res;
  }

  /**
   * When a compass reading is returned, it needs to be handled as a floating point number. This
   * method handles the conversion between types.
   * 
   * @param s The compass reading from the calling member
   * @return A numeric representation of the angle returned from the compass
   */
  private double decodeDouble(String s) {
    return Double.parseDouble(s);
  }

  /**
   * Once a command and it's associated data has been encoded, it will need to be sent. Once a
   * response has been sent back as a result of an operation, it needs to be read and returned for
   * decoding.
   * 
   * @param s The command/data to send over the serial connection
   * @return The return from a command call; At current will either be a waypoint, a lidar return or
   *         a compass reading
   * @throws SerialPortException If the connection is lost, then this method can complete its task
   *         successfully
   */
  private String call(String s) throws SerialPortException {

    LOGGER.log(Level.INFO, "String to send:" + s);

    p.purgePort(SerialPort.PURGE_RXCLEAR | SerialPort.PURGE_TXCLEAR);

    try {
      Thread.sleep(1000);
    } catch (InterruptedException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

    p.writeString(s);

    LOGGER.log(Level.INFO, "sent");

    StringBuilder sb = new StringBuilder();

    try {
      Thread.sleep(1000);
    } catch (InterruptedException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }

    char incomingChar = (char) (p.readBytes(1))[0];

    LOGGER.log(Level.INFO, "successfully read first char");

    while (incomingChar != '\n') {
      sb.append(incomingChar);
      incomingChar = (char) (p.readBytes(1))[0];
    }

    return sb.toString();
  }

  /**
   * Before being sent, the command/data must first be formatted in a way that it can be sent, read
   * and interpreted at the other end of the connection.
   * 
   * @param c The command to perform
   * @param o Extra data to append. For the drive command, this is a waypoint, otherwise it is
   *        unused and null should be provided as the argument
   * @return The encoded command string ready for sending over the Serial connection
   */
  private String encode(Commands c, Object o) {
    StringBuilder sb = new StringBuilder();

    switch (c) {
      case COMPASS:
        LOGGER.log(Level.FINEST, "Adding COMPASS");
        sb.append("COMPASS");
        break;
      case DRIVE:
        LOGGER.log(Level.FINEST, "Adding DRIVE");
        sb.append("DRIVE");
        sb.append(';');
        sb.append(((Waypoint) o).getX());
        sb.append(';');
        sb.append(((Waypoint) o).getY());
        break;
      case L_SENSE:
        LOGGER.log(Level.FINEST, "Adding LSENSE");
        sb.append("LSENSE");
        break;
    }

    sb.append('\n');
    LOGGER.log(Level.INFO, sb.toString());

    return sb.toString();
  }

  /**
   * As part of the shutdown process, this method will close the Serial connection established
   * between the two devices.
   */
  public void closePipe() {
    try {
      p.closePort();
    } catch (SerialPortException e) {
      e.printStackTrace();
    }
  }
}

File: mcomp_C/communication/Pipe.cpp
/*
 * Pipe.cpp
 *
 *  Created on: 4 Apr 2018
 *      Author: Ryan Shoobert 15812407
 *
 *  Pipe Class
 *
 *  This class handles commands/data received from a sender and decodes them into
 *  operations. Currently commands exist for driving, retrieving lidar reads and
 *  Retrieving a reading from the compass. To extend, this could handle readings
 *  from different sensors and driving different kinds of motors.
 */

#include "Pipe.h"

#include "../config/robot_config.h"

Pipe::Pipe() {
}

/**
 * Receive Command Method
 *
 * Once a command has been read from the serial port and decoded, it will need
 * to be checked and if matches with a defined command, it will be executed and
 * the data that is returned will need to be encoded for returning to the sender.
 */
void Pipe::recieveCommand() {

  String* res = new String[3];
  Propulsion prop = Propulsion();
  LSensor lSen = LSensor();

  CONSOLE.begin(CONSOLERATE);
  CONSOLE.setTimeout(CONSOLETIMEOUT);

  res = decode(call());

  //check the data read and build a response if it contains any matching commands
//  if (res[0] == "COMPASS") {
//    writeString(encodeDouble(Compass().compassRead()));
//  }
  if (res[0] == "DRIVE") {
    Waypoint w = Waypoint(res[1].toDouble(), res[2].toDouble());
    writeString(encodeWaypoint(prop.Drive(w)));
  }
  if (res[0] == "LSENSE") {
    writeString(encodeLRead(lSen.sense()));
  }
}

/**
 * Call Method
 *
 * When a command is sent, it must be read in and returned for decoding.
 * Until then, as a robot will only act when instructed to do so, this
 * will act as a blocking call until input is received.
 */
String Pipe::call() {

  String sb = "";

  while (!CONSOLE.available()) {
  }

  char incomingChar = CONSOLE.read();
  sb.concat(incomingChar);

  while (incomingChar != '\n') {
    while (!CONSOLE.available()) {
    }

    incomingChar = CONSOLE.read();

    if (incomingChar != '\n') {
      sb.concat(incomingChar);
    }
  }
  return sb;
}

/**
 * Write String Method
 *
 * Owing to the fact that Arduino Serial.write can't take a string, this method
 * will handle the writing of each character in a string to the serial port.
 */
void Pipe::writeString(String s) {
  for (unsigned int i = 0; i < s.length(); i++) {
    CONSOLE.write(s.charAt(i));
  }
}

/**
 * Decode Method
 *
 * Once a command has been received from the sender, if it contains extra information,
 * Separated by a delimiter (in the case of drive an DRIVE;x;y) then this must be
 * separated into pieces. If not, then anything in the read data will be returned whole.
 */
String* Pipe::decode(String readData) {

  int delims[2];
  String* chunks = new String[3];

  delims[0] = readData.indexOf(';');
  delims[1] = readData.indexOf(';', delims[0]);

  if (delims[0] >= 0) {
    chunks[0] = readData.substring(0,delims[0]);
    chunks[1] = readData.substring(delims[0],delims[1]);
    chunks[2] = readData.substring(delims[1]);
  } else {
    chunks[0] = readData;
  }

  //return separate parts of the read string
  return chunks;
}

/**
 * Encode Double Method
 *
 * Prior to sending the response to the compass read command, the output of the
 * read must first be encoded into a format it can be sent and decoded at
 * the receiving end.
 */
String Pipe::encodeDouble(double d) {
  String sb = String(d, 2);
  sb.concat('\n');
  return sb;
}

/**
 * Encode Lidar Read Method
 *
 * As a result of the lsense command being issued/lidar read being taken, the data
 * obtained from the sensor needs to be returned. This will enable it to be added to the
 * map to aid pathfinding.
 *
 * As a read is 360 long, it needs to be iterated through and formatted in delimiter (;)
 * seperated chunks.
 *
 */
String Pipe::encodeLRead(Waypoint* reading) {
  String sb = "";
  int i;

  for (i = 0; i < 359; i++) {
    sb.concat(encodeInlineWaypoint(*(reading + i)));
  }

  sb.concat(encodeWaypoint(*(reading + i)));

  return sb;
}

/**
 * Encode Inline Waypoint Method
 *
 * As part of the encoding for a lidar read, this method takes a waypoint set in the read
 * and encodes it minus a trailing new line character.
 */
String Pipe::encodeInlineWaypoint(Waypoint w) {
  String sb = "";

  sb.concat(w.getX());
  sb.concat(';');
  sb.concat(w.getY());
  sb.concat(';');

  return sb;
}

/**
 * Encode Waypoint Method
 *
 * Prior to returning a waypoint as a result of a drive operation,
 * either the x, y for the current location or the destination waypoint will need to
 * be formatted prior to sending.
 */
String Pipe::encodeWaypoint(Waypoint w) {
  String sb = "";

  sb.concat(w.getX());
  sb.concat(';');
  sb.concat(w.getY());

  sb.concat('\n');

  return sb;
}

/**
 * Close Method
 *
 * When shutting down, this method can be called to terminate the established serial
 * connection.
 */
void Pipe::close() {
//  CONSOLE.end();
}

/**
 * Destructor for the Pipe class.
 */
Pipe::~Pipe() {
  // currently not implemented
}

